### PR TITLE
[ROCm] utilize PUBLIC vs PRIVATE linking to avoid incorrect dependencies

### PR DESCRIPTION
**Pull request by jeffdaily**
Fixes the build of projects that depend on torch, such as torchaudio.  Otherwise torchaudio will complain that gloo_hip is missing.

### CODE REVIEW

**Comment by jithunnair-amd** [comment]
cc @micmelesse

File: caffe2/contrib/CMakeLists.txt [hunk 1]
```diff
@@ -32,5 +32,7 @@ set(Caffe2_GPU_BINARY_SRCS ${Caffe2_GPU_BINARY_SRCS} PARENT_SCOPE)
 
 # HIP sources, include, test sources
 set(Caffe2_HIP_SRCS ${Caffe2_HIP_SRCS} PARENT_SCOPE)
-set(Caffe2_HIP_TEST_SRCS ${Caffe2_HIP_TEST_SRCS} PARENT_SCOPE)
 set(Caffe2_HIP_INCLUDE ${Caffe2_HIP_INCLUDE} PARENT_SCOPE)
+set(Caffe2_HIP_DEPENDENCY_LIBS ${Caffe2_HIP_DEPENDENCY_LIBS} PARENT_SCOPE)
+set(Caffe2_HIP_TEST_SRCS ${Caffe2_HIP_TEST_SRCS} PARENT_SCOPE)
+set(Caffe2_HIP_BINARY_SRCS ${Caffe2_HIP_BINARY_SRCS} PARENT_SCOPE)
```
Comment (jithunnair-amd) [inline caffe2/contrib/CMakeLists.txt:38]: Is this variable used somewhere?

Comment (jithunnair-amd) [reply caffe2/contrib/CMakeLists.txt:38]: I meant `Caffe2_HIP_BINARY_SRCS`

Comment (jeffdaily) [reply caffe2/contrib/CMakeLists.txt:38]: Not entirely sure.  Added here for parity with the GPU/CUDA statements from above in the same file.

Comment (jeffdaily) [reply caffe2/contrib/CMakeLists.txt]: Removed Caffe2_HIP_BINARY_SRCS var.

**Review by ezyang — APPROVED** [review]
okey dokey

**Comment by ezyang** [comment]
merge conflicted :(

**Comment by ezyang** [comment]
I can't maintainer-push to the branch, you're going to have to fix this yourself

```
To github.com:pytorch/pytorch.git
 ! [remote rejected]       rocm_cmake_dep_fix -> refs/pull/54727/head (deny updating a hidden ref)
error: failed to push some refs to 'github.com:pytorch/pytorch.git'
```

**Comment by jeffdaily** [comment]
@ezyang fixed the conflict.  Thanks.

**Pull request closed without merging**